qbraid/interface/qbraid_qasm3/tools.py

```diff
@@ -72,6 +73,127 @@ def _convert_to_contiguous_qasm3(qasmstr: QASMType, rev_qubits=False) -> QASMTyp
+    else:
+        try:
+            name = qubit_decl.split("]")[1].split(";")[0].strip()
+            size = int(qubit_decl.split("[")[1].split("]")[0].strip())
+        except IndexError as _:
+            name = qubit_decl.split("qubit")[1].split(";")[0].strip()
+            size = 1
+    return (name, size)
+
+
+def _remove_gate_definitions(qasm_str):
+    """This is required to account for the case when the gate
+     definition has an argument which is having same name as a
+     quantum register
+
+     now, if any gate is applied on this argument, it will be
+     interpreted as being applied on THE WHOLE register, when it is
+     only applied on the argument.
+
+    Example :
+
+    gate custom q1 {
+        x q1; // this is STILL DETECTED as a gate application on q1
+    }
+    qreg q1[4];
+    qreg q2[2];
+    custom q1[0];
+    cx q1[1], q2[1];
+
+    // Actual depth : 1
+    // Calculated depth : 2 (because of the gate definition)
+    """
+
+    gate_decls = [x.group() for x in re.finditer(r"(gate)(.*\n)*?\s*\}", qasm_str)]
+    for decl in gate_decls:
+        qasm_str = qasm_str.replace(decl, "")
+    return qasm_str
+
+
+def _get_unused_qubit_indices(qasm_str, register_list):
+    qasm_str = _remove_gate_definitions(qasm_str)
+    lines = qasm_str.splitlines()
+    gate_lines = [
+        s
+        for s in lines
+        if s.strip() and not s.strip().startswith(("OPENQASM", "include", "qreg", "qubit", "//"))
+    ]
+    unused_indices = {}
+    for qreg, size in register_list:
+        unused_indices[qreg] = set(range(size))
+
+        for line in gate_lines:
+            if qreg not in line:
+                continue
+            # either qubits or full register is referenced
+            used_indices = {int(x) for x in re.findall(rf"{qreg}\[(\d+)\]", line)}
+            if len(used_indices) > 0:
+                unused_indices[qreg] = unused_indices[qreg].difference(used_indices)
+            else:
+                # full register is referenced
+                unused_indices[qreg] = set()
+                break
+
+            if len(unused_indices[qreg]) == 0:
+                break
+
+    return unused_indices
+
+
+def convert_to_contiguous_qasm3(qasmstr: str) -> QASMType:
+    """Converts OpenQASM 3 string to contiguous qasm3 string with gate expansion"""
+
+    # SCOPE : no loops, no functions at the moment
+
+    # pylint: disable=import-outside-toplevel
+
+    # 1. Analyse the qasm3 string for qubits and bit declarations
+    qubit_list = _qasm_qubit_decl(qasmstr)
+    qreg_list = []
+
+    # 2. Identify which qubits are not used in the circuit
+    for qubit_decl in qubit_list:
+        qreg_list.append(_get_qreg(qubit_decl))
+
+    # 3. Add an identity gate for the unused qubits
+    qubit_indices = _get_unused_qubit_indices(qasmstr, qreg_list)
+    expansion_qasm = ""
+    for reg, indices in qubit_indices.items():
+        for index in indices:
+            expansion_qasm += f"i {reg}[{index}];\n"
```

Comment:
For the default behavior of the convert to contiguous function, we want to collapse any empty registers. Optionally, you can add the `expansion` keyword argument to the function definition, e.g.

https://github.com/qBraid/qBraid/blob/7a436d804929404bef5ec271b687ef16981f4be1/qbraid/interface/qbraid_cirq/tools.py#L139C69-L139C84

which, if set to `True`, would instead add the identity gates to the empty registers. I think that's what you're attempting to do here, but the behavior isn't quite correct. Using your example test case:

```python
from qbraid.interface.qbraid_braket.qasm import braket_from_qasm3
from qbraid.interface.qbraid_qasm3.tools import convert_to_contiguous_qasm3

qasm_test = """
OPENQASM 3.0;
gate custom q1, q2, q3{
    x q1;
    y q2;
    z q3;
}
qreg q1[2];
qubit[2] q2;
qubit[3] q3;
qubit q4;

x q1[0];
y q2[0];
z q3;
"""

circuit_in = braket_from_qasm3(qasm_test)
circuit_out = braket_from_qasm3(convert_to_contiguous_qasm3(qasm_test))
```
```python
>>> print(circuit_in)
T  : |0|
        
q0 : -X-
        
q2 : -Y-
        
q4 : -Z-
        
q5 : -Z-
        
q6 : -Z-

T  : |0|

>>> print(circuit_out)
T  : |0|
        
q0 : -X-
        
q1 : -I-
        
q2 : -Y-
        
q3 : -I-
        
q4 : -Z-
        
q5 : -Z-
        
q6 : -Z-
        
q7 : -I-

T  : |0|
```

This input circuit was 5 qubits, with two empty registers. For for contiguous expansion, it should end up as 7 qubits, but this output circuit is 8 qubits.

Reply:
Aren't there a total of 8 qubits? 
q1 -> 2 qubits
q2 -> 2 qubits
q3 -> 3 qubits 
q4 -> 1 qubit

Reply:
One question arises here @ryanhill1 , if a register is partially used do we redeclare it as a register with a size equal to the used qubits or just let it be? 
Example - 
```
qubit[3] q1;
qubit[2] q2;
cx q1[0], q2[0];
cx q1[1], q2[1];
```
should be changed to - 
```
qubit[2] q1; // changed size
qubit[2] q2;
cx q1[0], q2[0];
cx q1[1], q2[1];
```

Reply:
The approach I see is that if `expansion` is not set, we remove the unused quantum registers i.e. where no qubits were accessed.

Then, we return the expansion of the leftover registers (with the same size declarations of partially used registers, just `I` gates added to unused ones)

Reply:
Yeah that's correct. And yes, we would re-declare the register with the adjusted size in that case.